sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_responses.hpp

```diff
@@ -210,4 +191,103 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
 
     friend class ShareFileClient;
   };
+
+  class ListSharesPagedResponse : public PagedResponse<ListSharesPagedResponse> {
```

Comment:
Yeah, it's a less common usage of template and inheritance. The only other time i've seen this is [std::enable_shared_from_this](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this)

Reply:
It's idiomatic enough to have a Wikipedia entry: https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern.

Reply:
> So the base class must be careful to avoid using the not-yet-complete derived class outside of function bodies.

What would that bad coding practice look like? I am trying to understand the types of mistakes/pitfalls i.e. things not to do in the base class to guard against this issue.

Reply:
> What would that bad coding practice look like? I am trying to understand the types of mistakes/pitfalls i.e. things not to do in the base class to guard against this issue.

Direct references to members of the (as yet incomplete) derived class outside of function bodies in the base class will result in a compile error. It's possible to workaround this restriction with even more template trickery, but it's best to keep the design simple (as we did with `PagedResponse`) to avoid the need in the first place. (My threshold hold for acceptable "template trickery" in the Azure C++ SDK design is very low; we don't want to end up with a design that only a small percentage of C++ experts understand.)